File: python_backend/artifact/subattr_distribution.py
from typing import Dict, List

from python_backend.artifact.artifact_attrs import ArtifactAttrs
from python_backend.artifact.rating import p_subattr_combination, artifact_remaining_enhance, \
    artifact_rating_expectation
from python_backend.artifact.weighted_attrs import WeightedAttrs
from python_backend.consts.terminology.artifact_consts import ArtifactEnum, ALL_4_ATTR_COMB, \
    P_3_SUBATTRS_ART_DOMAIN, P_4_SUBATTRS_ART_DOMAIN, MAX_NUM_ATTRS, ALL_3_ATTR_COMB
from python_backend.consts.terminology.attribute_consts import AttributeEnum, LIST_SUBATTR_RATIO, \
    AVG_SUBATTR_RATIO


def _scale_distribtion():
    distribution = [{0: 1}]
    len_ratios = len(LIST_SUBATTR_RATIO)
    for i in range(1, 11):  # FUTURE_TODO: magic number (maximum total of 9 subattrs enhancements)
        prev = distribution[i - 1]
        curr = {}
        for scale in prev:
            for ratio in LIST_SUBATTR_RATIO:
                key = scale + ratio
                curr[key] = curr.setdefault(key, 0) + prev[scale] * 1 / len_ratios
        distribution.append(curr)
    return distribution


def _lvl0_distribution():
    distribution = [{"data": [], "p": 1, "num": 0}]
    len_list = len(LIST_SUBATTR_RATIO)
    for i in range(1, 5):
        num = i
        p = 1 / len_list ** num
        list = []

        def recur(carry_list: List[float], recur_times: int):
            if recur_times == 0:
                list.append(carry_list.copy())
                return
            for j in LIST_SUBATTR_RATIO:
                carry_list.append(j)
                recur(carry_list, recur_times - 1)
                carry_list.pop()

        recur([], i)
        distribution.append({
            "data": list,
            "p": p,
            "num": num,
        })
    return distribution


_SCALE_DISTRIBUTION = _scale_distribtion()
_LVL0_DISTRIBUTION = _lvl0_distribution()


def scale_distribution():
    return _SCALE_DISTRIBUTION


class _LSDResult:
    def _max_value(self, case):
        score = 0
        for each in case:  # each[0]: value, each[2] num of scale
            score += each[0] * each[2]
        return score

    def _min_value(self, case):
        score = 0
        for each in case:  # each[0]: value, each[2] num of scale
            score += each[0] * each[2] * 0.7
        return score

    def _exp_case(self, case):
        score = 0
        for each in case:  # each[0]: value, each[1] weight
            score += each[0] * each[1] * AVG_SUBATTR_RATIO * (
                    self._remaining_enhance / MAX_NUM_ATTRS)
        return score

    def __init__(self, result: dict, lvl: int):
        self._lvl = lvl
        self._remaining_enhance = artifact_remaining_enhance(self._lvl)
        self._sorted_result = list(
            sorted(result.items(), reverse=True, key=lambda x: self._max_value(x[0])))
        self._max_values = []
        self._min_values = []
        self._exp_values = []
        for case in self._sorted_result:
            self._max_values.append(self._max_value(case[0]))
            self._min_values.append(self._min_value(case[0]))
            self._exp_values.append(self._exp_case(case[0]))
        self._len_result = len(self._sorted_result)

    @staticmethod
    def _p_in_actual_distribution(case, score: float):
        """
        :param case:
        :param score:
        :return: the probability that the case has score not smaller than the given score
        """
        len_case = len(case)
        p = []  # the probabilities that the case has score greater than the given score

        def recur(i, carried_score, carried_p):
            if i == len_case:
                if carried_score >= score:
                    p.append(carried_p)
                return
            value = case[i][0]
            enhances = case[i][2]
            distribution = _SCALE_DISTRIBUTION[enhances]
            for scale in distribution:
                recur(i + 1, carried_score + value * scale, carried_p * distribution[scale])
            pass

        recur(0, 0, 1)
        return sum(sorted(p))

    def p_score_greater(self, score: float, with_exp=False) -> float:
        """
        :return: the probability that a random artifact has score not lower than the provided one
        """
        adjusted_score = score - 0.001  # to correct floating point minor errors
        p = []  # the probabilities that the given score is not greater than a random artifact's
        for i in range(self._len_result):
            case = self._sorted_result[i][0]
            p_case = self._sorted_result[i][1]
            max_value = self._max_values[i]
            min_value = self._min_values[i]
            exp_value: float
            if with_exp:
                exp_value = self._exp_values[i]
            else:
                exp_value = 0
            score_diff_exp = adjusted_score - exp_value
            if max_value >= score_diff_exp:
                if min_value >= score_diff_exp:
                    p.append(p_case)
                else:
                    p.append(
                        p_case * self._p_in_actual_distribution(case, score_diff_exp))
        return sum(sorted(p))


def leveled_subattrs_distribution(mainattr: AttributeEnum, lvl: int,
                                  weights: WeightedAttrs) -> _LSDResult:
    """
    Only work for artifacts above (including) level 4.

    :param mainattr: the main attribute of the artifact
    :param lvl: the level of the artifact
    :return: a result class _LSDResult which can provide more information by calling its methods
    """
    if lvl < 4:
        print("[!] Invalid request for leveled_subattrs_distribution: lvl too low")
        return _LSDResult({}, -1)
    weighted_subattrs = ArtifactEnum.FLOWER.subattr_weights_readonly().copy()
    weighted_subattrs.pop(mainattr, None)
    curr_layer: dict = {}
    next_layer: dict = {}
    curr_level_covers: int = 7
    # Initialize for the sub-attrs of a random level 4 artifact
    for comb in ALL_4_ATTR_COMB:
        if mainattr not in comb:  # if this is a valid combination
            p_comb = p_subattr_combination(weighted_subattrs, comb)
            # initialize attributes
            attrs = []  # type is list[tuple[weight, scale]]
            attrs_count = {}
            for attr in comb:
                value = weights.get(attr)
                if value > 0:  # only record if attribute has value
                    attrs_count[value] = attrs_count.setdefault(value, 0) + 1
            for weight in attrs_count:
                attrs.append((weight, attrs_count[weight], attrs_count[weight]))
            attrs.sort(reverse=True)
            if len(attrs) > 0:
                key = tuple(attrs)
                # now attrs is in a situation of an originally 3-attr artifact
                curr_layer[key] = curr_layer.setdefault(key, 0) + P_3_SUBATTRS_ART_DOMAIN * p_comb
                # for original 4-attr artifact
                not_stronger_count = MAX_NUM_ATTRS
                len_attrs = len(attrs)
                # Calculate probability that weighted attribute is enhanced
                for i in range(len_attrs):
                    original_tuple = attrs[i]
                    weight = original_tuple[1]
                    attrs[i] = (original_tuple[0], original_tuple[1], original_tuple[2] + 1)
                    key = tuple(attrs)
                    p_chosen = weight / MAX_NUM_ATTRS
                    curr_layer[key] = curr_layer.setdefault(key, 0) \
                                      + P_4_SUBATTRS_ART_DOMAIN * p_comb * p_chosen
                    not_stronger_count -= weight
                    attrs[i] = original_tuple
                # Calculate probability that unweighted attribute is enhanced
                key = tuple(attrs)
                p_chosen = not_stronger_count / MAX_NUM_ATTRS
                curr_layer[key] = curr_layer.setdefault(key, 0) \
                                  + P_4_SUBATTRS_ART_DOMAIN * p_comb * p_chosen

    # Compute the sub-attrs for level 8+ artifacts if needed
    while curr_level_covers < lvl:
        for situation in curr_layer:
            not_stronger_count = MAX_NUM_ATTRS
            list_situation = list(situation)
            len_attrs = len(situation)
            p_situation = curr_layer[situation]
            # Calculate probability that weighted attribute is enhanced
            for i in range(len_attrs):
                original_tuple = list_situation[i]
                value = situation[i][0]
                weight = situation[i][1]
                scale = situation[i][2]
                list_situation[i] = (value, weight, scale + 1)
                key = tuple(list_situation)
                p_chosen = weight / MAX_NUM_ATTRS
                next_layer[key] = next_layer.setdefault(key, 0) + p_situation * p_chosen
                not_stronger_count -= weight
                list_situation[i] = original_tuple
            # Calculate probability that unweighted attribute is enhanced
            key = tuple(list_situation)
            p_chosen = not_stronger_count / MAX_NUM_ATTRS
            next_layer[key] = next_layer.setdefault(key, 0) + p_situation * p_chosen
        curr_layer = next_layer
        next_layer = {}
        curr_level_covers += 4

    return _LSDResult(curr_layer, lvl)


def relative_rarity_compare_subattrs(artifact_type: ArtifactEnum, mainattr: AttributeEnum,
                                     lvl: int,
                                     attrs: ArtifactAttrs, weights: WeightedAttrs) -> float:
    """
    Currently only work for level 0-3 artifacts.
    This is an approximation, which differs little from the actual value, but enough for reference.

    :param artifact_type:
    :param mainattr:
    :param lvl:
    :param attrs:
    :param weights:
    :return:
    """

    this_rating = artifact_rating_expectation(artifact_type, mainattr, lvl, attrs, weights)
    subattrs_except_main: Dict[AttributeEnum, float] = artifact_type.subattr_weights_readonly().copy()
    subattrs_except_main.pop(mainattr, None)

    def calculate_p(combs):
        p_list = []
        for comb in combs:
            if mainattr not in comb:
                p_comb = p_subattr_combination(subattrs_except_main, comb)
                comb_attrs = ArtifactAttrs()
                tmp_attrs = {}
                for attr in comb:
                    tmp_attrs[attr] = 0
                comb_attrs.set(tmp_attrs)
                # comb_attrs.set_avg(set(comb))
                comb_avg_rating = artifact_rating_expectation(artifact_type, mainattr, 0,
                                                              comb_attrs,
                                                              weights)
                max_base_rating = sum(
                    [LIST_SUBATTR_RATIO[-1] * weights.get(attr) for attr in comb]
                )
                if this_rating > comb_avg_rating + max_base_rating:
                    p_list.append(p_comb)
                    continue
                len_comb = len(comb)
                distribution = _LVL0_DISTRIBUTION[len_comb]
                for case in distribution["data"]:
                    base_rating = 0
                    for i in range(len_comb):
                        base_rating += case[i] * weights.get(comb[i])
                    if this_rating > comb_avg_rating + base_rating:
                        p_list.append(p_comb * distribution["p"])
        return p_list

    p_3: List[float] = calculate_p(ALL_3_ATTR_COMB)
    p_4: List[float] = calculate_p(ALL_4_ATTR_COMB)
    p_3.sort()
    p_4.sort()
    p = P_3_SUBATTRS_ART_DOMAIN * sum(p_3) + P_4_SUBATTRS_ART_DOMAIN * sum(p_4)
    return 1 - p


# if __name__ == '__main__':
#     start = time.time()
#     k = leveled_subattrs_distribution(
#         AttributeEnum.ATK_FLAT, 4, WeightedAttrs.crit_atk_er_em_plan()
#     ).p_score_greater(5.5)
#     end = time.time()
#     print("probability:", k)
#     print("time taken:", end - start)


if __name__ == '__main__':
    a_attrs = ArtifactAttrs().set_avg({
        AttributeEnum.CRIT_RATE,
        AttributeEnum.ATK_FLAT,
        AttributeEnum.ATK_PCT,
        # AttributeEnum.ATK_FLAT,
        AttributeEnum.ER,
        # AttributeEnum.EM,
    })
    a = relative_rarity_compare_subattrs(ArtifactEnum.FLOWER, AttributeEnum.HP_FLAT, 0, a_attrs
                                         , WeightedAttrs.crit_atk_plan())
    print(a)

File: react-native_frontend/Dottore/app/backend/artifact/subattr_distribution.js
import {
  ALL_3_ATTR_COMB,
  ALL_4_ATTR_COMB,
  ArtifactEnum,
  MAX_NUM_ATTRS,
  P_3_SUBATTRS_ART_DOMAIN,
  P_4_SUBATTRS_ART_DOMAIN,
} from "../consts/artifact_consts";
import { AVG_SUBATTR_RATIO, LIST_SUBATTR_RATIO } from "../consts/attribute_consts";
import { sum_array } from "../helpers/array_helper";
import { copy_object, get_or, set_if_exist } from "../helpers/object_helper";
import { ArtifactAttrs } from "./artifact_attrs";
import {
  artifact_rating_expectation,
  artifact_remaining_enhance,
  p_subattr_combination,
} from "./rating";
import { WeightedAttrs } from "./weighted_attrs";

/**
 * @returns {Array<Object<number, number>>} Array[Object[scale, p]]
 */
function scale_distribution() {
  const distribution = [{ 0: 1 }];
  len_ratios = LIST_SUBATTR_RATIO.length;
  for (let i = 1; i < 11; i++) {
    const prev = distribution[i - 1];
    const curr = {};
    Object.keys(prev).forEach((scale) => {
      LIST_SUBATTR_RATIO.forEach((ratio) => {
        const key = (Number(scale) + ratio).toFixed(1);
        curr[key] = get_or(curr, key, 0) + prev[scale] * (1 / len_ratios);
      });
    });
    distribution.push(curr);
  }
  return distribution;
}

/**
 * @returns {Array<Object>} Array<Object<.data<Array<Array<number>>>, .p, .num>>
 */
function lvl0_distribution() {
  distribution = [{ data: [], p: 1, num: 0 }];
  len_list = LIST_SUBATTR_RATIO.length;
  for (let i = 1; i < 5; i++) {
    const num = i;
    const p = (1 / len_list) ** num;
    const list = [];

    /**
     * @param {Array<number>} carry_list
     * @param {number} recur_times
     * @returns
     */
    function recur(carry_list, recur_times) {
      if (recur_times === 0) {
        list.push([...carry_list]);
        return;
      }
      LIST_SUBATTR_RATIO.forEach((j) => {
        carry_list.push(j);
        recur(carry_list, recur_times - 1);
        carry_list.pop();
      });
    }

    recur([], i);
    distribution.push({
      data: list,
      p,
      num,
    });
  }
  return distribution;
}

export const SCALE_DISTRIBUTION = scale_distribution();
export const LVL0_DISTRIBUTION = lvl0_distribution();

/**
 * @param {Array<number>} arr
 * @returns {string}
 */
function encode(arr) {
  return arr.map((item) => item.toString()).join(":");
}

/**
 * @param {string} s
 * @returns {Array<number>}
 */
function decode(s) {
  return s.split(":").map((val) => Number(val));
}

/**
 * @param {Array<Array<number>>} arr
 * @returns {string}
 */
function encodeArr(arr) {
  return arr.map(encode).join(";");
}

/**
 * @param {string} s
 * @returns {Array<Array<number>>}
 */
function decodeArr(s) {
  return s.split(";").map(decode);
}

class LSDResult {
  #LVL;
  #REMAINING_ENHANCE;
  #SORTED_RESULT;
  #MAX_VALUES;
  #MIN_VALUES;
  #EXP_VALUES;
  #LEN_RESULT;
  /**
   * @param {Array<Array<number>>} casee
   */
  private_max_value(casee) {
    let score = 0;
    casee.forEach((each) => {
      score += each[0] * each[2];
    });
    return score;
  }

  /**
   * @param {Array<Array<number>>} casee
   */
  private_min_value(casee) {
    let score = 0;
    casee.forEach((each) => {
      score += each[0] * each[2] * 0.7;
    });
    return score;
  }

  /**
   * @param {Array<Array<number>>} casee
   */
  private_exp_value(casee) {
    let score = 0;
    casee.forEach((each) => {
      score += each[0] * each[1] * AVG_SUBATTR_RATIO * (this.#REMAINING_ENHANCE / MAX_NUM_ATTRS);
    });
    return score;
  }

  /**
   * @param {Object<string, number>} result
   * @param {number} lvl
   */
  constructor(result, lvl) {
    this.#LVL = lvl;
    this.#REMAINING_ENHANCE = artifact_remaining_enhance(this.#LVL);
    const tmp_arr = Object.keys(result).map((key) => [decodeArr(key), result[key]]);
    tmp_arr.sort((a, b) => this.private_max_value(b[0]) - this.private_max_value(a[0]));
    this.#SORTED_RESULT = tmp_arr;
    this.#MAX_VALUES = [];
    this.#MIN_VALUES = [];
    this.#EXP_VALUES = [];
    this.#SORTED_RESULT.forEach((item) => {
      this.#MAX_VALUES.push(this.private_max_value(item[0]));
      this.#MIN_VALUES.push(this.private_min_value(item[0]));
      this.#EXP_VALUES.push(this.private_exp_value(item[0]));
    });
    this.#LEN_RESULT = this.#SORTED_RESULT.length;
  }

  /**
   * @param {Array<Array<number>>} casee
   * @param {number} score
   * @returns {number}
   */
  private_p_in_actual_distribution(casee, score) {
    const len_casee = casee.length;
    const p = [];

    function recur(i, carried_score, carried_p) {
      if (i === len_casee) {
        if (carried_score >= score) {
          p.push(carried_p);
        }
        return;
      }
      const value = casee[i][0];
      const enhances = casee[i][2];
      const distribution = SCALE_DISTRIBUTION[enhances];
      Object.keys(distribution).forEach((scale) => {
        recur(i + 1, carried_score + value * Number(scale), carried_p * distribution[scale]);
      });
    }

    recur(0, 0, 1);
    p.sort();
    return p.reduce((acc, nxt) => acc + nxt, 0);
  }

  /**
   * @param {number} score
   * @param {boolean} [withExp]
   * @returns {number}
   */
  p_score_greater(score, withExp = false) {
    const adjusted_score = score - 0.001;
    const p = [];
    for (let i = 0; i < this.#LEN_RESULT; i++) {
      const casee = this.#SORTED_RESULT[i][0];
      const p_casee = this.#SORTED_RESULT[i][1];
      const max_value = this.#MAX_VALUES[i];
      const min_value = this.#MIN_VALUES[i];
      let exp_value = 0;
      if (withExp) {
        exp_value = this.#EXP_VALUES[i];
      }
      const score_diff_exp = adjusted_score - exp_value;
      if (max_value >= score_diff_exp) {
        if (min_value >= score_diff_exp) {
          p.push(p_casee);
        } else {
          p.push(p_casee * this.private_p_in_actual_distribution(casee, score_diff_exp));
        }
      }
    }
    p.sort();
    return p.reduce((acc, nxt) => acc + nxt, 0);
  }
}

/**
 * @param {string} mainattr AttributeEnum
 * @param {number} lvl
 * @param {WeightedAttrs} weights
 * @returns {LSDResult}
 */
export function leveled_subattrs_distribution(mainattr, lvl, weights) {
  if (lvl < 4) {
    return new LSDResult({}, -1); // TODO: empty
  }
  const weighted_subattrs = copy_object(ArtifactEnum.FLOWER.subattr_weights_readonly);
  set_if_exist(weighted_subattrs, mainattr, 0);
  let curr_layer = {};
  let next_layer = {};
  let curr_level_covers = 7;
  // for Array<AttributeEnum.key> in ...
  ALL_4_ATTR_COMB.forEach((comb) => {
    if (!(mainattr in comb)) {
      const p_comb = p_subattr_combination(weighted_subattrs, comb);
      const attrs = [];
      const attrs_count = {};
      // for AttributeEnum.key in ...
      comb.forEach((attr) => {
        const value = weights.get(attr);
        if (value > 0) {
          attrs_count[value] = get_or(attrs_count, value, 0) + 1;
        }
      });
      // for number in ...
      Object.keys(attrs_count).forEach((weight) => {
        weight = Number(weight);
        attrs.push([weight, attrs_count[weight], attrs_count[weight]]);
      });
      attrs.sort((a, b) => b[0] - a[0]);
      if (attrs.length > 0) {
        let key = encodeArr(attrs);
        curr_layer[key] = get_or(curr_layer, key, 0) + P_3_SUBATTRS_ART_DOMAIN * p_comb;
        let not_stronger_count = MAX_NUM_ATTRS;
        const len_attrs = attrs.length;
        for (let i = 0; i < len_attrs; i++) {
          const original_tuple = attrs[i];
          const weight = original_tuple[1];
          attrs[i] = [original_tuple[0], original_tuple[1], original_tuple[2] + 1];
          key = encodeArr(attrs);
          const p_chosen = weight / MAX_NUM_ATTRS;
          curr_layer[key] =
            get_or(curr_layer, key, 0) + P_4_SUBATTRS_ART_DOMAIN * p_comb * p_chosen;
          not_stronger_count -= weight;
          attrs[i] = original_tuple;
        }
        key = encodeArr(attrs);
        const p_chosen = not_stronger_count / MAX_NUM_ATTRS;
        curr_layer[key] = get_or(curr_layer, key, 0) + P_4_SUBATTRS_ART_DOMAIN * p_comb * p_chosen;
      }
    }
  });

  while (curr_level_covers < lvl) {
    // for Array<Array<number>>.toString() in ...
    Object.keys(curr_layer).forEach((situation) => {
      let not_stronger_count = MAX_NUM_ATTRS;
      const arr_situation = decodeArr(situation);
      const len_attrs = arr_situation.length;
      const p_situation = curr_layer[situation];
      for (let i = 0; i < len_attrs; i++) {
        const original_tuple = arr_situation[i];
        const value = original_tuple[0];
        const weight = original_tuple[1];
        const scale = original_tuple[2];
        arr_situation[i] = [value, weight, scale + 1];
        const key = encodeArr(arr_situation);
        const p_chosen = weight / MAX_NUM_ATTRS;
        next_layer[key] = get_or(next_layer, key, 0) + p_situation * p_chosen;
        not_stronger_count -= weight;
        arr_situation[i] = original_tuple;
      }
      const key = encodeArr(arr_situation);
      const p_chosen = not_stronger_count / MAX_NUM_ATTRS;
      next_layer[key] = get_or(next_layer, key, 0) + p_situation * p_chosen;
    });
    curr_layer = next_layer;
    next_layer = {};
    curr_level_covers += 4;
  }

  // console.log("=========")
  // Object.keys(curr_layer).forEach((key) => {
  //   console.log(key, curr_layer[key])
  // })
  return new LSDResult(curr_layer, lvl);
}

/**
 * @param {string} artifact_type ArtifactEnum
 * @param {string} mainattr AttributeEnum
 * @param {number} lvl
 * @param {ArtifactAttrs} attrs
 * @param {WeightedAttrs} weights
 * @returns
 */
export function relative_rarity_compare_subattrs(artifact_type, mainattr, lvl, attrs, weights) {
  const this_rating = artifact_rating_expectation(artifact_type, mainattr, lvl, attrs, weights);
  const subattrs_except_main = copy_object(ArtifactEnum[artifact_type].subattr_weights_readonly);
  set_if_exist(subattrs_except_main, mainattr, 0);

  /**
   * @param {Array<Array>} combs
   */
  function calculate_p(combs) {
    const p_list = [];
    combs.forEach((comb) => {
      if (!(mainattr in comb)) {
        const p_comb = p_subattr_combination(subattrs_except_main, comb);
        const comb_attrs = new ArtifactAttrs();
        const tmp_attrs = {};
        comb.forEach((attr) => {
          tmp_attrs[attr] = 0;
        });
        comb_attrs.set(tmp_attrs);
        const comb_avg_rating = artifact_rating_expectation(
          artifact_type,
          mainattr,
          0,
          comb_attrs,
          weights
        );
        const max_base_rating =
          comb.map((attr) => weights.get(attr)).reduce((acc, nxt) => acc + nxt, 0) *
          LIST_SUBATTR_RATIO[LIST_SUBATTR_RATIO.length - 1];
        if (this_rating > comb_avg_rating + max_base_rating) {
          p_list.push(p_comb);
          return;
        }
        const len_comb = comb.length;
        const distribution = LVL0_DISTRIBUTION[len_comb];
        distribution.data.forEach((casee) => {
          let base_rating = 0;
          for (let i = 0; i < len_comb; i++) {
            base_rating += casee[i] * weights.get(comb[i]);
          }
          if (this_rating > comb_avg_rating + base_rating) {
            p_list.push(p_comb * distribution.p);
          }
        });
      }
    });
    return p_list;
  }

  const p_3 = calculate_p(ALL_3_ATTR_COMB);
  const p_4 = calculate_p(ALL_4_ATTR_COMB);
  p_3.sort();
  p_4.sort();
  p = P_3_SUBATTRS_ART_DOMAIN * sum_array(p_3) + P_4_SUBATTRS_ART_DOMAIN * sum_array(p_4);
  return 1 - p;
}
